vllm/engine/multiprocessing/mp_llm_engine.py

```diff
@@ -0,0 +1,259 @@
+            `parallel_config.worker_use_ray`.
+        engine_use_ray: Whether to make LLMEngine a Ray actor. If so, the
+            async frontend will be executed in a separate process as the
+            model workers.
+        async_engine_args: AsyncLLMEngine args
+        log_requests: Whether to log the requests.
+    """
+
+    _engine_class: Type[LLMEngine] = LLMEngine
+
+    def __init__(self,
+                 worker_use_ray: bool,
+                 engine_use_ray: bool,
+                 *args,
+                 ipc_path: str,
+                 log_requests: bool = True,
+                 **kwargs) -> None:
+
+        if engine_use_ray:
+            raise NotImplementedError("Not yet supported.")
+
+        self.worker_use_ray = worker_use_ray
+        self.engine_use_ray = engine_use_ray
+        self.log_requests = log_requests
+        self.engine = self._init_engine(*args, **kwargs)
+
+        self.ctx = zmq.Context()  # type: ignore[attr-defined]
+
+        # Receive input from the client.
+        self.input_socket = self.ctx.socket(zmq.constants.PULL)
+        self.input_socket.bind(f"{ipc_path}_input_socket")
+
+        # Send output stream back to client.
+        self.output_socket = self.ctx.socket(zmq.constants.PUSH)
+        self.output_socket.bind(f"{ipc_path}_output_socket")
+
+        # Send health status back to client.
+        self.health_socket = self.ctx.socket(zmq.constants.PUSH)
+        self.health_socket.bind(f"{ipc_path}_health_socket")
+
+        # IPC path for the data socket.
+        self.data_ipc_path = f"{ipc_path}_data_socket"
+
+    @classmethod
+    def from_engine_args(cls, engine_args: AsyncEngineArgs,
+                         usage_context: UsageContext, ipc_path: str):
+        """Creates an MPLLMEngine from the engine arguments."""
+
+        engine_config = engine_args.create_engine_config()
+
+        if engine_args.engine_use_ray:
+            from vllm.executor import ray_utils
+            ray_utils.assert_ray_available()
+
+        # TODO: better abstraction?
+        executor_class = AsyncLLMEngine._get_executor_cls(engine_config)
+
+        return cls(
+            executor_class.uses_ray,
+            engine_args.engine_use_ray,
+            **engine_config.to_dict(),
+            executor_class=executor_class,
+            log_requests=not engine_args.disable_log_requests,
+            log_stats=not engine_args.disable_log_stats,
+            usage_context=usage_context,
+            ipc_path=ipc_path,
+        )
+
+    def cleanup(self):
+        """Cleanup zeromq state on shutdown."""
+        self.input_socket.close()
+        self.output_socket.close()
+        self.ctx.destroy(linger=0)
+        del self.engine
+
+    def _init_engine(self, *args,
+                     **kwargs) -> Union[LLMEngine, "ray.ObjectRef"]:
+        """Initialize the LLMEngine"""
+
+        if not self.engine_use_ray:
+            engine_class = self._engine_class
+        elif self.worker_use_ray:
```

Comment:
what is `worker_use_ray`?

Reply:
it is the same as `--distributed_executor_backend=ray`